File: walk_route.go
// walk_route.go
package main

import (
	"encoding/json"
	"flag"
	"fmt"
	"io"
	"net/http"
	"os"
	"strconv"
	"strings"
)

type Waypoint struct {
	Name string  `json:"name"`
	Lat  float64 `json:"lat"`
	Lng  float64 `json:"lng"`
}

type RouteData struct {
	Waypoints []Waypoint `json:"waypoints"`
	Speed     float64    `json:"speed"`
	Weight    float64    `json:"weight"`
}

type RoutePlanner struct {
	Data RouteData `json:"data"`
	File string
}

func NewRoutePlanner(file string) *RoutePlanner {
	rp := &RoutePlanner{File: file}
	rp.load()
	if rp.Data.Speed == 0 {
		rp.Data.Speed = 5.0
	}
	if rp.Data.Weight == 0 {
		rp.Data.Weight = 70.0
	}
	return rp
}

func (rp *RoutePlanner) load() {
	data, err := os.ReadFile(rp.File)
	if err != nil {
		return
	}
	json.Unmarshal(data, &rp.Data)
}

func (rp *RoutePlanner) save() {
	data, _ := json.MarshalIndent(rp.Data, "", "  ")
	os.WriteFile(rp.File, data, 0644)
}

func geocode(query string) (float64, float64, error) {
	url := fmt.Sprintf("https://nominatim.openstreetmap.org/search?q=%s&format=json&limit=1", query)
	req, _ := http.NewRequest("GET", url, nil)
	req.Header.Set("User-Agent", "WalkRoutePlanner/1.0")
	client := &http.Client{}
	resp, err := client.Do(req)
	if err != nil {
		return 0, 0, err
	}
	defer resp.Body.Close()
	body, _ := io.ReadAll(resp.Body)
	var data []map[string]string
	if err := json.Unmarshal(body, &data); err != nil {
		return 0, 0, err
	}
	if len(data) == 0 {
		return 0, 0, fmt.Errorf("not found")
	}
	lat, _ := strconv.ParseFloat(data[0]["lat"], 64)
	lng, _ := strconv.ParseFloat(data[0]["lon"], 64)
	return lat, lng, nil
}

func (rp *RoutePlanner) add(location string) {
	if strings.Contains(location, ",") {
		parts := strings.Split(location, ",")
		if len(parts) == 2 {
			lat, err1 := strconv.ParseFloat(strings.TrimSpace(parts[0]), 64)
			lng, err2 := strconv.ParseFloat(strings.TrimSpace(parts[1]), 64)
			if err1 == nil && err2 == nil {
				name := fmt.Sprintf("(%.4f,%.4f)", lat, lng)
				rp.Data.Waypoints = append(rp.Data.Waypoints, Waypoint{Name: name, Lat: lat, Lng: lng})
				rp.save()
				fmt.Printf("✅ Added waypoint: %s\n", name)
				return
			}
		}
	}
	lat, lng, err := geocode(location)
	if err != nil {
		fmt.Printf("❌ Could not geocode '%s': %v\n", location, err)
		return
	}
	rp.Data.Waypoints = append(rp.Data.Waypoints, Waypoint{Name: location, Lat: lat, Lng: lng})
	rp.save()
	fmt.Printf("✅ Added waypoint: %s (%.4f, %.4f)\n", location, lat, lng)
}

func (rp *RoutePlanner) list() {
	if len(rp.Data.Waypoints) == 0 {
		fmt.Println("No waypoints.")
		return
	}
	fmt.Println("\n📍 Waypoints:")
	for i, w := range rp.Data.Waypoints {
		fmt.Printf("%d: %s (%.4f, %.4f)\n", i+1, w.Name, w.Lat, w.Lng)
	}
}

func (rp *RoutePlanner) remove(index int) {
	if index < 1 || index > len(rp.Data.Waypoints) {
		fmt.Printf("❌ Invalid index. Must be between 1 and %d\n", len(rp.Data.Waypoints))
		return
	}
	removed := rp.Data.Waypoints[index-1]
	rp.Data.Waypoints = append(rp.Data.Waypoints[:index-1], rp.Data.Waypoints[index:]...)
	rp.save()
	fmt.Printf("✅ Removed waypoint: %s\n", removed.Name)
}

func (rp *RoutePlanner) clear() {
	rp.Data.Waypoints = []Waypoint{}
	rp.save()
	fmt.Println("✅ All waypoints cleared.")
}

func (rp *RoutePlanner) speed(speed float64) {
	rp.Data.Speed = speed
	rp.save()
	fmt.Printf("✅ Walking speed set to %.1f km/h\n", speed)
}

func (rp *RoutePlanner) route() {
	if len(rp.Data.Waypoints) < 2 {
		fmt.Println("Need at least 2 waypoints to calculate a route.")
		return
	}
	coords := ""
	for i, w := range rp.Data.Waypoints {
		if i > 0 {
			coords += ";"
		}
		coords += fmt.Sprintf("%f,%f", w.Lng, w.Lat)
	}
	url := fmt.Sprintf("http://router.project-osrm.org/route/v1/walking/%s?overview=full&geometries=polyline&steps=true", coords)
	resp, err := http.Get(url)
	if err != nil {
		fmt.Printf("❌ Error fetching route: %v\n", err)
		return
	}
	defer resp.Body.Close()
	body, _ := io.ReadAll(resp.Body)
	var data map[string]interface{}
	if err := json.Unmarshal(body, &data); err != nil {
		fmt.Println("❌ Invalid response from routing service.")
		return
	}
	routes, ok := data["routes"].([]interface{})
	if !ok || len(routes) == 0 {
		fmt.Println("❌ No route found.")
		return
	}
	route := routes[0].(map[string]interface{})
	distanceKm := route["distance"].(float64) / 1000
	durationMin := route["duration"].(float64) / 60
	// Estimate elevation gain
	elevation := distanceKm * 12
	calories := distanceKm * 55 * (rp.Data.Weight / 70.0)

	fmt.Printf("\n🚶 Walking Route:\n")
	fmt.Printf("Total distance: %.2f km\n", distanceKm)
	fmt.Printf("Total duration: %.1f min (at %.1f km/h)\n", durationMin, rp.Data.Speed)
	fmt.Printf("Elevation gain: ~%.0f m\n", elevation)
	fmt.Printf("Calories burned: ~%.0f kcal\n", calories)

	legs, ok := route["legs"].([]interface{})
	if ok && len(legs) > 0 {
		fmt.Println("\nTurn-by-turn:")
		stepNum := 1
		for _, leg := range legs {
			steps, _ := leg.(map[string]interface{})["steps"].([]interface{})
			for _, s := range steps {
				st := s.(map[string]interface{})
				maneuver := st["maneuver"].(map[string]interface{})
				instruction := maneuver["instruction"].(string)
				dist := st["distance"].(float64) / 1000
				fmt.Printf("%d. %s (%.2f km)\n", stepNum, instruction, dist)
				stepNum++
			}
		}
	}
}

func main() {
	if len(os.Args) < 2 {
		fmt.Println("Usage: walk_route <command> [options]")
		return
	}
	rp := NewRoutePlanner("walk_route.json")
	cmd := os.Args[1]
	switch cmd {
	case "add":
		if len(os.Args) < 3 {
			fmt.Println("add <location>")
			return
		}
		rp.add(os.Args[2])
	case "list":
		rp.list()
	case "route":
		rp.route()
	case "remove":
		if len(os.Args) < 3 {
			fmt.Println("remove <index>")
			return
		}
		idx, _ := strconv.Atoi(os.Args[2])
		rp.remove(idx)
	case "clear":
		rp.clear()
	case "speed":
		if len(os.Args) < 3 {
			fmt.Println("speed <kmh>")
			return
		}
		speed, _ := strconv.ParseFloat(os.Args[2], 64)
		rp.speed(speed)
	default:
		fmt.Println("Unknown command. Use add, list, route, remove, clear, speed.")
	}
}

File: WalkRoute.java
// WalkRoute.java
import java.io.*;
import java.net.*;
import java.nio.file.*;
import java.util.*;
import com.google.gson.*;

class Waypoint {
    String name;
    double lat;
    double lng;

    Waypoint(String name, double lat, double lng) {
        this.name = name;
        this.lat = lat;
        this.lng = lng;
    }
}

class RouteData {
    List<Waypoint> waypoints = new ArrayList<>();
    double speed = 5.0;
    double weight = 70.0;
}

class RoutePlanner {
    private RouteData data = new RouteData();
    private final String dataFile = "walk_route.json";
    private final Gson gson = new GsonBuilder().setPrettyPrinting().create();

    public RoutePlanner() { load(); }

    private void load() {
        try {
            Path path = Paths.get(dataFile);
            if (Files.exists(path)) {
                String json = new String(Files.readAllBytes(path));
                data = gson.fromJson(json, RouteData.class);
            }
        } catch (Exception e) {}
    }

    private void save() {
        try {
            Files.write(Paths.get(dataFile), gson.toJson(data).getBytes());
        } catch (Exception e) {}
    }

    private double[] geocode(String query) throws Exception {
        String url = "https://nominatim.openstreetmap.org/search?q=" + URLEncoder.encode(query, "UTF-8") + "&format=json&limit=1";
        HttpURLConnection conn = (HttpURLConnection) new URL(url).openConnection();
        conn.setRequestMethod("GET");
        conn.setRequestProperty("User-Agent", "WalkRoutePlanner/1.0");
        BufferedReader reader = new BufferedReader(new InputStreamReader(conn.getInputStream()));
        StringBuilder sb = new StringBuilder();
        String line;
        while ((line = reader.readLine()) != null) sb.append(line);
        reader.close();
        JsonArray arr = gson.fromJson(sb.toString(), JsonArray.class);
        if (arr.size() == 0) return null;
        JsonObject obj = arr.get(0).getAsJsonObject();
        double lat = obj.get("lat").getAsDouble();
        double lng = obj.get("lon").getAsDouble();
        return new double[]{lat, lng};
    }

    public void add(String location) throws Exception {
        if (location.contains(",")) {
            String[] parts = location.split(",");
            if (parts.length == 2) {
                try {
                    double lat = Double.parseDouble(parts[0].trim());
                    double lng = Double.parseDouble(parts[1].trim());
                    String name = String.format("(%.4f,%.4f)", lat, lng);
                    data.waypoints.add(new Waypoint(name, lat, lng));
                    save();
                    System.out.println("✅ Added waypoint: " + name);
                    return;
                } catch (NumberFormatException e) {}
            }
        }
        double[] coords = geocode(location);
        if (coords == null) {
            System.out.println("❌ Could not geocode '" + location + "'");
            return;
        }
        data.waypoints.add(new Waypoint(location, coords[0], coords[1]));
        save();
        System.out.printf("✅ Added waypoint: %s (%.4f, %.4f)%n", location, coords[0], coords[1]);
    }

    public void list() {
        if (data.waypoints.isEmpty()) {
            System.out.println("No waypoints.");
            return;
        }
        System.out.println("\n📍 Waypoints:");
        for (int i = 0; i < data.waypoints.size(); i++) {
            Waypoint w = data.waypoints.get(i);
            System.out.printf("%d: %s (%.4f, %.4f)%n", i+1, w.name, w.lat, w.lng);
        }
    }

    public void remove(int index) {
        if (index < 1 || index > data.waypoints.size()) {
            System.out.printf("❌ Invalid index. Must be between 1 and %d%n", data.waypoints.size());
            return;
        }
        Waypoint removed = data.waypoints.remove(index-1);
        save();
        System.out.println("✅ Removed waypoint: " + removed.name);
    }

    public void clear() {
        data.waypoints.clear();
        save();
        System.out.println("✅ All waypoints cleared.");
    }

    public void speed(double speed) {
        data.speed = speed;
        save();
        System.out.printf("✅ Walking speed set to %.1f km/h%n", speed);
    }

    public void route() throws Exception {
        if (data.waypoints.size() < 2) {
            System.out.println("Need at least 2 waypoints to calculate a route.");
            return;
        }
        StringBuilder coords = new StringBuilder();
        for (Waypoint w : data.waypoints) {
            if (coords.length() > 0) coords.append(";");
            coords.append(w.lng).append(",").append(w.lat);
        }
        String url = "http://router.project-osrm.org/route/v1/walking/" + coords + "?overview=full&geometries=polyline&steps=true";
        HttpURLConnection conn = (HttpURLConnection) new URL(url).openConnection();
        conn.setRequestMethod("GET");
        BufferedReader reader = new BufferedReader(new InputStreamReader(conn.getInputStream()));
        StringBuilder sb = new StringBuilder();
        String line;
        while ((line = reader.readLine()) != null) sb.append(line);
        reader.close();
        JsonObject obj = gson.fromJson(sb.toString(), JsonObject.class);
        if (!obj.has("routes") || obj.get("routes").getAsJsonArray().size() == 0) {
            System.out.println("❌ No route found.");
            return;
        }
        JsonObject route = obj.get("routes").getAsJsonArray().get(0).getAsJsonObject();
        double distanceKm = route.get("distance").getAsDouble() / 1000;
        double durationMin = route.get("duration").getAsDouble() / 60;
        double elevation = distanceKm * 12;
        double calories = distanceKm * 55 * (data.weight / 70.0);

        System.out.println("\n🚶 Walking Route:");
        System.out.printf("Total distance: %.2f km%n", distanceKm);
        System.out.printf("Total duration: %.1f min (at %.1f km/h)%n", durationMin, data.speed);
        System.out.printf("Elevation gain: ~%.0f m%n", elevation);
        System.out.printf("Calories burned: ~%.0f kcal%n", calories);

        if (route.has("legs")) {
            System.out.println("\nTurn-by-turn:");
            int stepNum = 1;
            for (JsonElement legElem : route.get("legs").getAsJsonArray()) {
                JsonObject leg = legElem.getAsJsonObject();
                if (leg.has("steps")) {
                    for (JsonElement stepElem : leg.get("steps").getAsJsonArray()) {
                        JsonObject step = stepElem.getAsJsonObject();
                        JsonObject maneuver = step.get("maneuver").getAsJsonObject();
                        String instruction = maneuver.get("instruction").getAsString();
                        double dist = step.get("distance").getAsDouble() / 1000;
                        System.out.printf("%d. %s (%.2f km)%n", stepNum, instruction, dist);
                        stepNum++;
                    }
                }
            }
        }
    }

    public static void main(String[] args) throws Exception {
        if (args.length < 1) {
            System.out.println("Usage: WalkRoute <command> [options]");
            return;
        }
        RoutePlanner planner = new RoutePlanner();
        String cmd = args[0];
        switch (cmd) {
            case "add":
                if (args.length < 2) { System.out.println("add <location>"); return; }
                planner.add(args[1]);
                break;
            case "list":
                planner.list();
                break;
            case "route":
                planner.route();
                break;
            case "remove":
                if (args.length < 2) { System.out.println("remove <index>"); return; }
                planner.remove(Integer.parseInt(args[1]));
                break;
            case "clear":
                planner.clear();
                break;
            case "speed":
                if (args.length < 2) { System.out.println("speed <kmh>"); return; }
                planner.speed(Double.parseDouble(args[1]));
                break;
            default:
                System.out.println("Unknown command. Use add, list, route, remove, clear, speed.");
        }
    }
}
